core/src/main/java/org/jboss/jandex/DotName.java

```diff
@@ -326,14 +353,27 @@ public String toString() {
      */
     public String toString(char delim) {
         if (componentized) {
-            StringBuilder builder = new StringBuilder();
+            if (delim == '.' && startsWithJava()) {
+                return JAVA_STRINGS.computeIfAbsent(this, COMPONENTIZED_TO_STRING);
+            }
+            StringBuilder builder = new StringBuilder(stringLength());
             buildString(delim, builder);
             return builder.toString();
         } else {
             return delim == '.' ? local : local.replace('.', delim);
         }
     }
 
+    private int stringLength() {
+        int length = local.length();
+        DotName prefix = this.prefix;
+        while (prefix != null) {
+            length += 1 + prefix.local.length();
+            prefix = prefix.prefix;
+        }
+        return length;
```

Comment:
FWIW, in my initial patch, I used the same pattern as for `buildString()` so maybe you want to drop recursion from there too as you apparently weren't fan of it?